hazelcast-sql/src/main/java/com/hazelcast/sql/impl/calcite/validate/SqlNodeUtil.java

```diff
@@ -0,0 +1,112 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.sql.impl.calcite.validate;
+
+import com.hazelcast.sql.SqlErrorCode;
+import com.hazelcast.sql.impl.QueryException;
+import com.hazelcast.sql.impl.type.converter.StringConverter;
+import org.apache.calcite.runtime.CalciteContextException;
+import org.apache.calcite.sql.SqlDynamicParam;
+import org.apache.calcite.sql.SqlKind;
+import org.apache.calcite.sql.SqlLiteral;
+import org.apache.calcite.sql.SqlNode;
+import org.apache.calcite.sql.SqlUtil;
+import org.apache.calcite.sql.type.SqlTypeName;
+
+import java.math.BigDecimal;
+
+import static org.apache.calcite.sql.type.SqlTypeName.APPROX_TYPES;
+import static org.apache.calcite.sql.type.SqlTypeName.CHAR_TYPES;
+import static org.apache.calcite.sql.type.SqlTypeName.DECIMAL;
+import static org.apache.calcite.sql.type.SqlTypeName.EXACT_TYPES;
+import static org.apache.calcite.util.Static.RESOURCE;
+
+/**
+ * Utility methods to work with {@link SqlNode}s.
+ */
+public final class SqlNodeUtil {
+
+    private SqlNodeUtil() {
+    }
+
+    /**
+     * @return {@code true} if the given node is a {@linkplain SqlDynamicParam
+     * dynamic parameter}, {@code false} otherwise.
+     */
+    public static boolean isParameter(SqlNode node) {
+        return node.getKind() == SqlKind.DYNAMIC_PARAM;
+    }
+
+    /**
+     * @return {@code true} if the given node is a {@linkplain SqlLiteral literal},
+     * {@code false} otherwise.
+     */
+    public static boolean isLiteral(SqlNode node) {
+        return node.getKind() == SqlKind.LITERAL;
+    }
+
+    /**
+     * Obtains a numeric value of the given node if it's a numeric or string
+     * {@linkplain SqlLiteral literal}.
+     * <p>
+     * If the literal represents an exact value (see {@link
+     * SqlTypeName#EXACT_TYPES}), the obtained numeric value is {@link BigDecimal}.
+     * Otherwise, if the literal represents an approximate value (see {@link
+     * SqlTypeName#APPROX_TYPES}), the obtained numeric value is {@link Double}.
+     *
+     * @param node the node to obtain the numeric value of.
+     * @return the obtained numeric value or {@code null} if the given node is
+     * not a numeric or string literal.
+     * @throws CalciteContextException if the given node is a string literal
+     *                                 that doesn't have a valid numeric
+     *                                 representation.
+     */
+    public static Number numericValue(SqlNode node) {
```

Comment:
There is one very subtle thing with literal conversions. Consider the following expressions and produced result types. The concrete function is not that important here, because the problem appears to be in the inference mechanics:
```
SELECT ABS(1) ... // DECIMAL
SELECT ABS(`1`) ... // DECIMAL due to inference VARCHAR -> DECIMAL
SELECT ABS(1E0) ... // DOUBLE
SELECT ABS(`1E0`) ... // DECIMAL due to inference path VARCHAR -> DECIMAL
```
The last result could be surprising to users - somehow an inexact expression is converted to the exact form. I do not say that this is definitely a bug, but ideally we should have a consistent behavior.

Reply:
I don't see inference path for `SELECT ABS(`1E0`)` in `SqlNodeUtil.numericValue` leading to `DECIMAL`. Most likely no custom type coercion/inference is even performed for `ABS`, what you see is the standard Calcite behavior, it coerces `VARCHAR` to `DECIMAL` in numeric contexts. Feel free to provide a fix in math functions PR.

Reply:
Sorry, I do not quite understand what sort of fix do you mean. What I see is that the problem is in our code: the literal is resolved correctly to `DOUBLE`. But before that, the `CAST` to `DECIMAL` is added in `HazelcastTypeCoercion#coerceOperandType`. Why do we add this cast in case of literal?

Reply:
To clarify - I'd like to understand how to use the basis we built for new functions. It seems that it is not enough to just declare the new function with the proper parameter in the `HazelcastSqlOperatorTable`. What is supposed to be the proper process of doing it?

Reply:
Perhaps the problem is in `HazelcastSqlValidator#deriveLiteralType` - when a literal with single quotes is passed there, the resulting literal type is VARCHAR. It seems that if we resolve the type properly here to DECIMAL/DOUBLE (in addition to the existing logic), the problem will go away.

Reply:
What you see is the standard Calcite behavior, it coerces `VARCHAR` to `DECIMAL` in numeric contexts, `ABS` expects numeric value, so Calcite inserts `CAST('1E0' AS DECIMAL)`, so '1E0' is now interpreted as a numeric value, for which a proper type ('DOUBLE') is inferred since it represents an approximate numeric literal, but the cast is still there: `CAST('1E0':DOUBLE AS DECIMAL):DECIMAL`.

Not quite sure how the proposed fix of `HazelcastSqlValidator.deriveLiteralType` might work, basically all `VARCHAR` literals resembling numbers would receive numeric types: `LENGTH('1E0') -> LENGTH(CAST('1E0' AS DOUBLE))`, that should produce an error since there is no implicit conversion from `DOUBLE` to `VARCHAR` and there is no `LENGTH` overload accepting `DOUBLE`.

Most likely a proper fix would require some customization of `TypeCoercionImpl` behavior using `HazelcastTypeCoercion`. Not sure how the potential fix would look like at the moment, let's leave it for math functions PR.

Reply:
In the case of `ABS`, the target type is `NUMERIC`, while the operand type is `VARCHAR. The latter comes from our code `HazelcastSqlValidator.deriveLiteralType`. If I add additional logic, that returns DOUBLE for this literal, the test passes as expected: target type is NUMERIC, the literal type is DOUBLE -> final type is DOUBLE and no conversion is needed.

I think we should not postpone this to other expressions because this is not tied to math functions. It seems natural to have a consistent type derivation logic for literals implemented in the first place so that we do not need to adjust literals in every future expression separately.

Reply:
I tried to apply the fix, formally the expression worked, but now other things like `SELECT LENGTH('1e5')` show incorrect results. 
Yes, it seems that literal handling is context-dependent, but it is still an open question - do we understand how to handle literals properly for an arbitrary function? Is it supposed that we will write a custom handling for every function?

Reply:
Honestly, I feel that I lost in our coercion strategies and the way it works in general. I would propose to have a design document with a high-level explanation of how it works. Otherwise, we would end up with a code full of magic that is hard to change and maintain (I mean mainly the class `HazelcastTypeCoercion`).

Reply:
To clarify on this comment - this is not a blocker for merge. We may go without aligned literals semantics because literals are still work in progress (e.g. temporal ones), and specifically about quoted inexact literals - there is a workaround not to use quotes. 

Regarding design, I have a feeling that we do not have a clear picture of how expressions should work in general. We have some conversions and coercion here and there, but it is hard to see the design behind them. Nevertheless, I have a feeling that we already did much more than many other products wrt Calcite integration, and created a good basis. Ideally, we should come with a nice and consistent design in 4.2, because if there is a common-sense on how the system works that users could grasp easily, it will make user experience much better. To me, a good example is Postgres, where they explain functions through overloads, that could even be queried from the system table. We have a plenty of things to understand - how to deal with parameters, what exactly "implicit conversion" means (e.g. should `LENGTH(boolField)` work or not?), etc.

Therefore, marking this discussion as resolved for now, to avoid delays with merge.